scipy/optimize/_differentialevolution.py

```diff
@@ -394,84 +502,81 @@ def __init__(self, func, bounds, args=(),
 
         # default population initialization is a latin hypercube design, but
         # there are other population initializations possible.
+        if init not in ('latinhypercube', 'random'):
+            raise ValueError("The population initialization method must be one"
+                             "of 'latinhypercube' or 'random'")
+        self._init_population_method = init
         self.num_population_members = popsize * self.parameter_count
 
         self.population_shape = (self.num_population_members,
                                  self.parameter_count)
+        self._nfev = None
+        self._nit = None
+        self.result = None
+        self._status_flag = None
 
+    def __enter__(self):
         self._nfev = 0
-        if init == 'latinhypercube':
-            self.init_population_lhs()
-        elif init == 'random':
-            self.init_population_random()
-        else:
-            raise ValueError("The population initialization method must be one"
-                             "of 'latinhypercube' or 'random'")
-
-        self.disp = disp
-
-    def init_population_lhs(self):
-        """
-        Initializes the population with Latin Hypercube Sampling.
-        Latin Hypercube Sampling ensures that each parameter is uniformly
-        sampled over its range.
-        """
-        rng = self.random_number_generator
-
-        # Each parameter range needs to be sampled uniformly. The scaled
-        # parameter range ([0, 1)) needs to be split into
-        # `self.num_population_members` segments, each of which has the following
-        # size:
-        segsize = 1.0 / self.num_population_members
-
-        # Within each segment we sample from a uniform random distribution.
-        # We need to do this sampling for each parameter.
-        samples = (segsize * rng.random_sample(self.population_shape)
-
-        # Offset each segment to cover the entire parameter range [0, 1)
-                   + np.linspace(0., 1., self.num_population_members,
-                                 endpoint=False)[:, np.newaxis])
-
-        # Create an array for population of candidate solutions.
-        self.population = np.zeros_like(samples)
-
-        # Initialize population of candidate solutions by permutation of the
-        # random samples.
-        for j in range(self.parameter_count):
-            order = rng.permutation(range(self.num_population_members))
-            self.population[:, j] = samples[order, j]
-
-        # reset population energies
+        self._nit = 0
+        assert self._init_population_method in ('latinhypercube', 'random')
```

Comment:
better make this a proper ``if not ...: raise ValueError`` check - ``assert``s get stripped so we tend to avoid them.